### PR TITLE
 refactor(v-tabs): refactored `v-tabs` functionality

### LINKED ISSUE #1516: V-Tabs on touch devices are scrollable even with nothing to scroll
### Steps to reproduce
1. Create `v-tabs`
2. Use Dev tools to inspect on mobile device.

### Versions
Vuetify v0.15.0-beta.4
Latest Vue/Chrome

### What is expected ?
When there is no overflow, disable scrolling on touch devices.

### What is actually happening ?
Even with only one tab and `:scrollable='false'` you can swipe the tabs bar.

### Reproduction Link
https://codepen.io/anon/pen/aLZbBw

### LINKED ISSUE #1900: v-tabs with hyperlinks
I don't know if it's a feature request or a bug.

### Steps to reproduce
1. Go to the first example of the v-tab (https://vuetifyjs.com/components/tabs)
2. Edit in codepen
3. Change the `:href=` props to an hyperlink (`href="https://github.com"`)
4. When clicking on tab, the browser doesn't follow the hyperlink

### Versions
0.15.7

### What is expected ?
I should be redirected to `github.com`

### What is actually happening ?
The components search for a `<v-tabs-items>` where its `id` is `github.com`.

### Reproduction Link
https://codepen.io/anon/pen/EwZMJP

### Context

In many case, I have a server which already has a router. I would like to use the `v-tabs` component with this router and I don't want to use the client side router. Usually, the server send page like:

route /
```
<...>
  <v-tabs-items href="/" class="active"></v-tabs-item>
   <v-tabs-items href="/settings"></v-tabs-item>
</...>
```

route /settings
```
<...>
  <v-tabs-items href="/"></v-tabs-item>
  <v-tabs-items href="/settings" class="active"></v-tabs-item>
</...>
```

Moreover, hyperlinks work with a `v-btn` ( ex: `<v-btn href="http://google.com">` will redirect me to `google.com`. I would except the same behavior with v-tabs.

The generated markup with tabs looks good: 
```html
<li class="tabs__li">
  <a href="http://google.com" class="tabs__item tabs__item--active" style="position: relative;">
    Item 1
  </a>
</li>
```

Also in the doc, `to` and `href` have the same description. Maybe `to` should be used for `vue-router` and `href` for real links ?

Thank you very much for the amazing work you're doing :)

### LINKED ISSUE #2299: tabs item with fixed props inside the v-tabs not supporting >=10 length in every tabs-item?
<b>Versions</b>
Vue: 2.0
Vuetify: 0.16
OS: Ubuntu
Browsers: Google Chrome

<b>What is expected ?</b>
the `v-tabs-item` with `fixed` props inside the `v-tabs` should support it for each items

<b>What is actually happening ?</b>
tabs item with `fixed` props inside the `v-tabs`component is not supporting >=10 length in every `v-tabs-item`, but the other tabs that has >=10 chars length (not all items has >=10 character just 1/2 items only tht has it) is working.
i know i can put grow to fix it, but if i did tht, when i add more item, the right/left_arrow icon will be brake and the scroll bar will be shown underneath the slider

<b>Reproduction Link</b>
https://codepen.io/anon/pen/YrMmLa

### LINKED ISSUE #2335: Horizontal scrolling of overflowing content in tabs changes the tab
### Steps to reproduce
Create swipeable tabs with content wider than the tab width, scroll the tab content

### Versions
0.17

### What is expected ?
Content should be scrolled, tab should not change

### What is actually happening ?
Content is scrolled, later tab is changed

### Reproduction Link
Go to the API section https://dev.vuetifyjs.com/components/tabs in mobile and try to scroll the content to see the props description

### LINKED ISSUE #2530: [Feature Request] Make tabs match material spec better
The tabs do not currently match the material specification and are rather difficult to use on mobile and narrow browser windows. 

In particular, the side margins need to be looked at:
image

Also see the discussion in #1789 and #1749

https://material.io/guidelines/components/tabs.html#

yeah, I didn't use the template. Sue me

### LINKED ISSUE #2548: [Bug Report] Incorrect display of v-tabs-slider on narrow screens
### Versions and Environment
**Vuetify:** 0.17.1
**Vue:** 2.5.4
**Browsers:** Google Chrome
**OS:** Windows 10

### Steps to reproduce
When I'm at [http://localhost/incidents](http://localhost/incidents) and I update the page on a narrow screen (1250px), I get a problem with the moved v-tabs-slider [img](https://yadi.sk/i/UBluSEMt3PnYhY)

### Expected Behavior
Correct display of v-tabs-slider on all screen sizes

### Actual Behavior
Incorrect display of v-tabs-slider on some screen sizes

### Reproduction Link
[https://codepen.io/Neiromaster/pen/EbbLVq?&editors=101](https://codepen.io/Neiromaster/pen/EbbLVq?&editors=101)

### LINKED ISSUE #2592: [Bug Report] Problems with v-model active class when using router-view and v-tabs
### Versions and Environment
**Vuetify:** 0.17.2
**Vue:** 2.5.7
**Browsers:** Firefox 57.0
**OS:** Linux x86_64

### Steps to reproduce
- Create a tab template as in https://vuetifyjs.com/components/tabs
- Add VueRouter to the template
- Add :to="tab" inside v-for in v-tabs-item
- Add router-view inside v-tabs-content

### Expected Behavior
v-model="active" should get the active tab when using router-link (:to) when changing the component inside v-tabs-content.
v-tabs-slider should follow the new created (active) tab and move to the previous tab when the current tab is closed.

### Actual Behavior
The router-view changes the component correcly but the v-model="active" doesn't follow the selected tab when using router and v-tabs-slider doesn't move to the previous (and now active) tab when I close the current tab.

### Reproduction Link
[https://codepen.io/fredd_/pen/mqpgBR](https://codepen.io/fredd_/pen/mqpgBR)

### Other comments
Hi guys, I'm new to Vuejs and Vuetify.

What I'm trying to make is a dynamic tab that can be added when clicked in a link on navigation drawer and removed when clicked in the "x" inside tab label.

When I am not using the router to redirect the tab-content to the respective component, all works good, I can close the current tab and it will move to the previous one (except for the tab-slider which does not follow to the previous tab).

But when I try to use the router to load the components inside v-tabs-content then the v-model stops getting the selected active tab making the closeTab method not to work properly. If I click to close tab the active tab gets back to the first tab and not to the one before the current closed tab. This happens cuz I can't get the v-model to work properly inside the v-tabs when using router.

To test what I'm saying just remove ":to="item" from v-tabs-item and "router-view" from v-card. You will see that the "selected tab" text changes acordingly to the active tab, but it does not happen when using the router to change the view (including the content transition animation between tabs).

I hope I had explained correctly what is happening to me.

Sorry if I misspelled something, english is not my main language.

### LINKED ISSUE #2736: [Bug Report] v-tabs-slider disappears with v-if after unloading and reloading
### Versions and Environment
**Vuetify:** 0.16.9
**Vue:** 2.5.2
**Browsers:** Chrome 62.0.3202.94
**OS:** Mac OS 10.10.5

### Steps to reproduce
1. Click the checkmark button in the top right
2. Observe that the tabs now appear
3. Click on any of the tabs
4. Observe that the v-tabs-slider is visible
5. Click on the close button in the top right
6. Observe that the tabs now disappear
7. Click on the checkmark button in the top right
8. Observe that the tabs now appear again
9. Click on any of the tabs
10. Observe that the v-tabs-slider does not appear anymore

### Expected Behavior
If a v-tabs-bar is controlled by v-if, the v-tabs-slider should be visible when the v-tabs-bar is visible when the v-if condition changes from true to false and back to true.

### Actual Behavior
If a v-tabs-bar is controlled by v-if, the v-tabs-slider is only visible when the v-tabs-bar becomes visible the first time. When the v-if condition changes from true to false and back to true, the v-tabs-slider is no longer visible when the v-if condition becomes true the second time and onwards.

### Reproduction Link
[https://codepen.io/smanji/pen/qVGpLL](https://codepen.io/smanji/pen/qVGpLL)

### Other comments
This bug is only present with v-if. If changed to use v-show, this issue is not present. However, this is not a suitable workaround because with v-show, the height of the toolbar is always extended to include space for the v-tabs-bar even which v-show evaluates to false. This is especially undesirable for mobile due to the limited screen real estate.

### LINKED ISSUE #2749: Missing pagination controls on tabs
Pagination controls are not visible, not even in the official website. It works only for mobile devices
captura de pantalla 2017-12-07 a la s 12 29 51 p m

### LINKED ISSUE #2756: [Bug Report] Tabs don't have same size when wrapped in toolbar and the fixed prop is used.
### Versions and Environment
**Vuetify:** 0.17.4
**Vue:** 2.5.9
**Browsers:** Chrome 62.0.3202.94
**OS:** Mac OS 10.13.1

### Steps to reproduce
The link shows the actual versus the expected.

### Expected Behavior
The tabs should be the same size.

### Actual Behavior
The tabs are automatically sized according to content.

### Reproduction Link
[https://codepen.io/anon/pen/JOQwbr?&editors=101](https://codepen.io/anon/pen/JOQwbr?&editors=101)

### PR DESCRIPTION
**Pull request by johnleider**
## Information
This refactor is to bring the `v-tabs` family of components more in line with Material Design [spec](https://material.io/guidelines/components/tabs.html) while also adding and improving existing functionality. All prop based settings have been moved from `v-tabs` to `v-tabs-bar`. The reasoning for this is that all functionality relating to items of `v-tabs-bar` have been moved to that component. Converted all **tabs** components to use **registrable** mixin.

## New functionality
- Now supports **append-icon** and **prepend-icon**
- Added **align-with-title** prop that will line up `v-tabs-item`'s with the `v-toolbar-title` component. Caveats: Must be using a `v-btn` before `v-toolbar-title` and wrapped text in `v-tabs-item` will make it appear unaligned (this is how it actually looks in the spec. 
image
- Added **right** prop to align tab items to the right
- Now supports the **height** prop
- Updated **mobile-break-point** to match the proper spec default, (1280 - 16px)
- The **tabs** prop now sets the extension height of `v-toolbar` to match the default height of `v-tabs` (_48px_)

## Breaking changes
- All props with the exception of **value** on `v-tabs` should be moved to `v-tabs-bar`
- **fixed** prop renamed to **fixed-tabs** (now on `v-tabs-bar`)
- **icons** prop renamed to **icons-and-text** (now on `v-tabs-bar`)
- removed the **scrollable** prop
- will now show arrows on all breakpoints if overflow is detected. This can be removed using the **hide-arrows** prop. Paired with the vuetify **breakpoint** object, user can now conditionally do `<v-tabs-bar :hide-arrows="$vuetify.breakpoint.smAndDown">`.
- If there are no registered `v-tabs-content` components and no **value** is provided, `v-tabs` will no longer automatically select the first `v-tabs-item`. The user should either explicitly define a model or provide content. With this caveat, we can now track the active target when using router links.

## Issues
- fixes #2756
- fixes #2749
- fixes #2736
- fixes #2592
- fixes #2548
- fixes #2530
- fixes #2335
- fixes #2299
- fixes #1900
- fixes #1516

## Playground (WIP)
```vue
<template>
  <div id="app">
    <v-app id="inspire">
      <v-tabs centered>
          <v-tab to="/page1">a</v-tabs-item>
          <v-tab to="/page2">aaaaaaaaaaaaaaaaaaaaaaaaaaa</v-tabs-item>
      </v-tabs>
      <v-tabs fixed-tabs>
          <v-tab to="/page1">a</v-tabs-item>
          <v-tab to="/page2">aaaaaaaaaaaaaaaaaaaaaaaaaaa</v-tabs-item>
      </v-tabs>
      <v-toolbar>
        <v-toolbar-side-icon></v-toolbar-side-icon>
        <v-toolbar-title>Testing</v-toolbar-title>
        <v-tabs slot="extension" align-with-title height="64">
            <v-tab to="/page1">a</v-tabs-item>
            <v-tab to="/page2">aaaaaaaaaaaaaaaaaaaaaaaaaaa</v-tabs-item>
        </v-tabs>
      </v-toolbar>
    </v-app>
  </div>
</template>
```

### CODE REVIEW

File: src/stylus/components/_toolbar.styl [hunk 1]
```diff
@@ -120,14 +123,6 @@ theme(toolbar, "toolbar")
   &--dense
     toolbar-content(7px)
 
-.toolbar__extension
-  .toolbar__title
-    margin-left: 72px !important
-
-  .tabs__bar
-    align-self: flex-end
```
Comment (KaelWD) [inline src/stylus/components/_toolbar.styl]: Why was this removed? There's blank space at the bottom of the toolbar now.

**Comment by KaelWD** [comment]
Can you provide a summary of the breaking changes, and maybe some snippets to test? None of the examples from the documentation are working correctly any more.

File: src/stylus/components/_toolbar.styl (hunk 1 above)
Comment (KaelWD) [reply src/stylus/components/_toolbar.styl]: image
https://codepen.io/anon/pen/baBYLa?editors=1010

File: src/util/helpers.js [hunk 2]
```diff
@@ -129,6 +129,13 @@ const tagsToReplace = {
...
   '>': '&gt;'
 }
+
 export function escapeHTML (str) {
   return str.replace(/[&<>]/g, tag => tagsToReplace[tag] || tag)
 }
+
+export function filterChildren (array = [], tag) {
+  return array.filter(child => {
+    return child.componentOptions.Ctor.options.name === tag
```
Comment (KaelWD) [inline src/util/helpers.js]: Can't you just do `child.$options.name`?

File: src/components/VTabs/mixins/tabs-touch.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+          return { offset, index }
+        }
+        offset = newOffset
+      }
+
+      return null
+    },
+    onTouchStart (e) {
+      this.startX = this.scrollOffset + e.touchstartX
+      this.$refs.container.style.transition = 'none'
```
Comment (KaelWD) [inline src/components/VTabs/mixins/tabs-touch.js]: Set `will-change: transform` too

File: src/stylus/components/_tabs.styl [hunk 4]
```diff
@@ -3,159 +3,161 @@
...
+      margin-right: 40px
+
+  &--overflow      
+    .tabs__div
+      flex: 1 0 auto
+
+  &--align-with-title
+    .tabs__container > div:first-child,
+    .tabs__slider-wrapper + .tabs__div
+      margin-left: 56px
```
Comment (KaelWD) [inline src/stylus/components/_tabs.styl]: These can just be padding on `.tabs__container`

File: src/components/VTabs/mixins/tabs-touch.js (hunk 3 above)
Comment (johnleider) [reply src/components/VTabs/mixins/tabs-touch.js]: Use case?

File: src/stylus/components/_tabs.styl (hunk 4 above)
Comment (johnleider) [reply src/stylus/components/_tabs.styl]: done

File: src/components/VTabs/mixins/tabs-touch.js (hunk 3 above)
Comment (KaelWD) [reply src/components/VTabs/mixins/tabs-touch.js]: Performance. You're changing `translateX` without warning the browser about it first.

**Comment by KaelWD** [comment]
Couple of issues still:

Template: https://pastebin.com/uKt4APMn
- [ ] `color` still applies to the items too, classes should be added to `tabs__container` instead
- [ ] same goes for `dark`
- [ ] Arrows should be displayed on desktop resolutions when overflowing
- [ ] `align-with-title` should be the default on mobile when overflowing
- [ ] changing the model should scroll the new active tab into view

Template: https://pastebin.com/vpb1VkqJ
- [ ] It'd be cool if we could use `name` to identify tabs instead of `href` and `id`
- [ ] `align-with-title` should be 68px on desktop, 60px on mobile
  - [ ] Left toolbar button has too much padding (not really related to this PR, but it is an issue)
- [ ] ~~overflow isn't detected until screen resize (can't scroll without resizing the window first)~~
  - Works if you add `:href="'#' + item"` to the tabs, something to look at for the `name` thing
- [ ] scroll past the end is too far, it should be the same as the left padding instead of half the width

**Comment by KaelWD** [comment]
This is really, really bad. I don't know what's happening with `appendIconVisible`, but it's causing the entire component tree to re-render whenever `scrollOffset` changes. 
image

I think it has something to do with the render magic going on in the render functions (`if ((!this[`${direction}IconVisible`]`)

**Comment by KaelWD** [comment]
That resolved one reason for the re-renders, but there's still a pile of spaghetti to wade through to figure out the others.

File: src/components/VTabs/mixins/tabs-touch.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        const newOffset = offset + items[index].clientWidth
+        if (newOffset >= container.clientWidth) {
+          return { offset: this.scrollOffset - offset, index: index + 1 }
+        }
+        offset = newOffset
+      }
+
+      return { offset: 0, index: 0 }
+    },
+    newOffsetAppend (container, items, offset = this.scrollOffset) {
```
Comment (KaelWD) [inline src/components/VTabs/mixins/tabs-touch.js]: This gets called with integers, but then tries to access properties on them as if they were DOM nodes?

File: src/stylus/components/_tabs.styl [hunk 6]
```diff
@@ -2,160 +2,151 @@
...
 
 /* Theme */
-tabs($material)
-  .tabs__li
+tabs__container($material)
+  background-color: $material.cards
+
+  .icon--left,
+  .icon--right
+    color: $material.icons.inactive
```
Comment (jacekkarczmarczyk) [inline src/stylus/components/_tabs.styl]: Where are these styles used? I guess it was for append/prepend icons in 0.17, but looks like it's not necessary any longer

**Comment by jacekkarczmarczyk** [comment]
playground for 492e0dc525bbde5df3b87f5d81e37d7c92ba997a (omitting href/to/id)
https://gist.github.com/jacekkarczmarczyk/fe4f710f60e9e49a696d860b76ea9696
